File: 0940-fruit-into-baskets/0940-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer, Integer> mpp =new HashMap<>();
        int l=0, r=0, maxLen =0, n=fruits.length;

        while(r<n){
            mpp.put(fruits[r], mpp.getOrDefault(fruits[r],0)+1);

            if(mpp.size()>2){
                 mpp.put(fruits[l], mpp.getOrDefault(fruits[l],0)-1);
                 if(mpp.get(fruits[l])==0)mpp.remove(fruits[l]);
                 l++;
            }

            if(mpp.size()<=2)maxLen=Math.max(maxLen, (r-l+1));
            r++;
        }
        return maxLen;
        
    }
}

File: 0940-fruit-into-baskets/0940-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        uniqueFruits = {}

        l,r,res =0,0,0

        while r< len(fruits):
            if fruits[r] not in uniqueFruits:
                uniqueFruits[fruits[r]]=1
            else:    
                uniqueFruits[fruits[r]]+=1
          

            while l<r and len(uniqueFruits)>2:
                uniqueFruits[fruits[l]]-=1
                if uniqueFruits[fruits[l]]==0:
                    del uniqueFruits[fruits[l]]
                l+=1    
            
            res =max(res,(r-l+1))
            r+=1

        return  res    
        